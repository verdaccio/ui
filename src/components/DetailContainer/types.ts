import { ReactNode } from 'react';

export interface Props {
  children: ReactNode;
  open: boolean;
  onClose: () => void;
}
