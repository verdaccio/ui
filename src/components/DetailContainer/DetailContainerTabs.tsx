import React from 'react';
import styled from '@emotion/styled';
import { useTranslation } from 'react-i18next';

import { default as MuiTabs } from '../../muiComponents/Tabs';
import Tab from '../../muiComponents/Tab';

interface Props {
  onChange: (event, newValue) => void;
  tabPosition: number;
}

const DetailContainerTabs: React.FC<Props> = ({ tabPosition, onChange }) => {
  const { t } = useTranslation();

  return (
    <Tabs
      indicatorColor={'primary'}
      onChange={onChange}
      textColor={'primary'}
      value={tabPosition}
      variant={'fullWidth'}>
      <Tab data-testid={'readme-tab'} id={'readme-tab'} label={t('tab.readme')} />
      <Tab data-testid={'dependencies-tab'} id={'dependencies-tab'} label={t('tab.dependencies')} />
      <Tab data-testid={'versions-tab'} id={'versions-tab'} label={t('tab.versions')} />
      <Tab data-testid={'uplinks-tab'} id={'uplinks-tab'} label={t('tab.uplinks')} />
    </Tabs>
  );
};

export default DetailContainerTabs;

const Tabs = styled(MuiTabs)({
  marginBottom: 16,
});
