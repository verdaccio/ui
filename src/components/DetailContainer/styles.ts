import styled from 'react-emotion';

export const Content = styled('div')`
  && {
    padding: 15px;
  }
`;
