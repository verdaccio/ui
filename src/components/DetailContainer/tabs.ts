export enum TabPosition {
  README = 'readme',
  DEPENDENCIES = 'dependencies',
  VERSIONS = 'versions',
  UPLINKS = 'uplinks',
}
