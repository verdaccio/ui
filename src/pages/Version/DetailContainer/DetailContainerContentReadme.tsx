import React from 'react';

import { preventXSS } from 'verdaccio-ui/utils/sec-utils';

import Readme from './Readme';

interface Props {
  description?: string;
}

const DetailContainerContentReadme: React.FC<Props> = ({ description }) => {
  if (!description) return null;
  const encodedReadme = preventXSS(description);
  return <Readme description={encodedReadme} />;
};

export default DetailContainerContentReadme;
