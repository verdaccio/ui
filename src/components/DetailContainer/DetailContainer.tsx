import React, { Component, ReactElement, Fragment } from 'react';

import { DetailContextConsumer, VersionPageConsumerProps } from '../../pages/version/Version';
import Readme from '../Readme';
import Versions from '../Versions';
import { preventXSS } from '../../utils/sec-utils';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';
import { Content } from './styles';
import Dependencies from '../Dependencies';
import UpLinks from '../UpLinks';

interface DetailContainerState {
  tabPosition: number;
}

class DetailContainer extends Component<any, DetailContainerState> {
  public state = {
    tabPosition: 0,
  };

  public render(): ReactElement<HTMLElement> {
    return (
      <DetailContextConsumer>
        {context => {
          return this.renderTabs(context as VersionPageConsumerProps);
        }}
      </DetailContextConsumer>
    );
  }

  private handleChange = (event: any, tabPosition: number) => {
    event.preventDefault();
    this.setState({ tabPosition });
  };

  private renderListTabs(): React.ReactElement<HTMLElement> {
    return (
      <Fragment>
        <Tab id={'readme-tab'} label={'Readme'} />
        <Tab id={'dependencies-tab'} label={'Dependencies'} />
        <Tab id={'versions-tab'} label={'Versions'} />
        <Tab id={'uplinks-tab'} label={'Uplinks'} />
      </Fragment>
    );
  }

  private renderTabs = ({ readMe }) => {
    const { tabPosition } = this.state;

    return (
      <Fragment>
        <Content>
          <Tabs indicatorColor={'primary'} onChange={this.handleChange} textColor={'primary'} value={tabPosition} variant={'fullWidth'}>
            {this.renderListTabs()}
          </Tabs>
          <br />
          {tabPosition === 0 && this.renderReadme(readMe)}
          {tabPosition === 1 && <Dependencies />}
          {tabPosition === 2 && <Versions />}
          {tabPosition === 3 && <UpLinks />}
        </Content>
      </Fragment>
    );
  };

  private renderReadme = (readMe: string): ReactElement<HTMLElement> => {
    const encodedReadme = preventXSS(readMe);

    return <Readme description={encodedReadme} />;
  };
}

export default DetailContainer;
