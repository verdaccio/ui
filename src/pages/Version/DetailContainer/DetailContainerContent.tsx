import React from 'react';

import Dependencies from './Dependencies';
import DetailContainerContentReadme from './DetailContainerContentReadme';
import { TabPosition } from './tabs';
import UpLinks from './UpLinks';
import Versions from './Versions';

interface Props {
  tabPosition: TabPosition;
  readDescription?: string;
}

const DetailContainerContent: React.FC<Props> = ({ tabPosition, readDescription }) => {
  switch (tabPosition) {
    case TabPosition.README:
      return <DetailContainerContentReadme description={readDescription} />;
    case TabPosition.UPLINKS:
      return <UpLinks />;
    case TabPosition.VERSIONS:
      return <Versions />;
    case TabPosition.DEPENDENCIES:
      return <Dependencies />;
    default:
      return null;
  }
};

export default DetailContainerContent;
