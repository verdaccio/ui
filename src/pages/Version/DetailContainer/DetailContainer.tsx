import React, { useState, useContext } from 'react';

import Box from 'verdaccio-ui/components/Box';

import { DetailContext } from '../context';

import Deprecated from './Deprecated';
import DetailContainerContent from './DetailContainerContent';
import DetailContainerTabs from './DetailContainerTabs';
import { TabPosition } from './tabs';

const DetailContainer: React.FC = () => {
  const tabs = Object.values(TabPosition);
  const [tabPosition, setTabPosition] = useState(0);
  const detailContext = useContext(DetailContext);
  const { readMe, packageMeta } = detailContext;

  const handleChange = (event: React.ChangeEvent<{}>, newValue: number) => {
    setTabPosition(newValue);
  };

  return (
    <Box component="div" display="flex" flexDirection="column" padding={2}>
      <DetailContainerTabs onChange={handleChange} tabPosition={tabPosition} />
      {packageMeta?.latest?.deprecated && <Deprecated message={packageMeta?.latest?.deprecated} />}
      <DetailContainerContent readDescription={readMe} tabPosition={tabs[tabPosition]} />
    </Box>
  );
};

export default DetailContainer;
