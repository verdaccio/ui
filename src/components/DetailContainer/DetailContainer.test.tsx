import React from 'react';
import { render } from '@testing-library/react';

import DetailContainer from './DetailContainer';

describe('DetailContainer', () => {
  test('renders correctly', () => {
    const { container } = render(<DetailContainer />);
    expect(container.firstChild).toMatchSnapshot();
  });
  test.todo('should test click on tabs');
});
