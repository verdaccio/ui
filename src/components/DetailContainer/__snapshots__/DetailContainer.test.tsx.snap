// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailContainer renders correctly 1`] = `
.emotion-0 {
  margin-bottom: 16px;
}

<div
  class="MuiBox-root MuiBox-root-2"
>
  <div
    class="MuiTabs-root emotion-0 emotion-1"
  >
    <div
      class="MuiTabs-scroller MuiTabs-fixed"
      style="overflow: hidden;"
    >
      <div
        class="MuiTabs-flexContainer"
        role="tablist"
      >
        <button
          aria-selected="true"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected MuiTab-fullWidth"
          data-testid="readme-tab"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            Readme
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-fullWidth"
          data-testid="dependencies-tab"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            Dependencies
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-fullWidth"
          data-testid="versions-tab"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            Versions
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          aria-selected="false"
          class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary MuiTab-fullWidth"
          data-testid="uplinks-tab"
          role="tab"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiTab-wrapper"
          >
            Uplinks
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      <span
        class="PrivateTabIndicator-root-27 PrivateTabIndicator-colorPrimary-28 MuiTabs-indicator"
        style="left: 0px; width: 0px;"
      />
    </div>
  </div>
</div>
`;
